plugins/inputs/huawei_routers_telemetry/huawei_routers_telemetry.go

```diff
@@ -0,0 +1,219 @@
+		err = proto.Unmarshal(gpbkv.Content, sensorMsg)
+		if err != nil {
+			fmt.Println("Sensor Error: ", err.Error())
+			return nil, err
+		}
+		fields, vals := huawei_sensorPath.SearchKey(gpbkv, msg.GetSensorPath())
+		tags := make(map[string]string, len(fields)+3)
+		tags["source"] = msg.GetNodeIdStr()
+		tags["subscription"] = msg.GetSubscriptionIdStr()
+		tags["path"] = msg.GetSensorPath()
+		// Search for Tags
+		for i := 0; i < len(fields); i++ {
+			tags = huawei_sensorPath.AppendTags(fields[i], vals[i], tags, msg.GetSensorPath())
+		}
+		// Create Metrics
+		for i := 0; i < len(fields); i++ {
+			huawei_sensorPath.CreateMetrics(grouper, tags, timestamp, msg.GetSensorPath(), fields[i], vals[i])
+		}
+	}
+	return grouper, nil
+}
+
+/*
+  Listen UDP packets and call the telemetryDecoder.
+*/
+func (h *packetSocketListener) listen() {
+	buf := make([]byte, 64*1024) // 64kb - maximum size of IP packet
+	for {
+		n, _, err := h.ReadFrom(buf)
+		if err != nil {
+			h.Log.Error("Unable to read buffer: %s", err.Error())
+			break
+		}
+
+		body, err := h.decoder.Decode(buf[:n])
+		if err != nil {
+			h.Log.Errorf("Unable to decode incoming packet: %s", err.Error())
+			continue
+		}
+		// Telemetry parsing over packet payload
+		grouper, err := HuaweiTelemetryDecoder(body)
+		if err != nil {
+			h.Log.Errorf("Unable to decode telemetry information: %s", err.Error())
+			break
+		}
+		for _, metric := range grouper.Metrics() {
+			h.AddMetric(metric)
+		}
+
+		if err != nil {
+			h.Log.Errorf("Unable to parse incoming packet: %s", err.Error())
+		}
+	}
+}
+
+type HuaweiRoutersTelemetry struct {
+	ServicePort     string        `toml:"service_port"`
+	ReadBufferSize  internal.Size `toml:"read_buffer_size"`
+	ContentEncoding string        `toml:"content_encoding"`
+	wg              sync.WaitGroup
+
+	Log telegraf.Logger `toml:"-"`
+
+	telegraf.Accumulator
+	io.Closer
+	decoder internal.ContentDecoder
+}
+
+func (h *HuaweiRoutersTelemetry) Description() string {
+	return "Huawei Telemetry UDP model input Plugin"
+}
+
+func (h *HuaweiRoutersTelemetry) SampleConfig() string {
+	return `
+  ## UDP Service Port to capture Telemetry
+  # service_port = "8080"
+
+`
+}
+
+func (h *HuaweiRoutersTelemetry) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (h *HuaweiRoutersTelemetry) Start(acc telegraf.Accumulator) error {
+	h.Accumulator = acc
+
+	var err error
+	h.decoder, err = internal.NewContentDecoder(h.ContentEncoding)
+	if err != nil {
+		return err
+	}
+
+	pc, err := udpListen("udp", ":"+h.ServicePort)
+	if err != nil {
+		return err
+	}
+
+	if h.ReadBufferSize.Size > 0 {
+		if srb, ok := pc.(setReadBufferer); ok {
+			srb.SetReadBuffer(int(h.ReadBufferSize.Size))
+		} else {
+			h.Log.Warnf("Unable to set read buffer on a %s socket", "udp")
+		}
+	}
+
+	h.Log.Infof("Listening on %s://%s", "udp", pc.LocalAddr())
+
+	psl := &packetSocketListener{
+		PacketConn:             pc,
+		HuaweiRoutersTelemetry: h,
+	}
```

Comment:
Can we please invert this and simply add the connection as a field to the `HuaweiRoutersTelemetry` structure. This is the way all other plugins go, so it is of least surprise.

Reply:
done